lib/kinesis_client.rb

```diff
@@ -82,21 +77,31 @@ def push_batch(batch)
 
     $logger.debug("Received #{resp} from #{@stream_name}")
 
-    return_message = {
-      failures: resp.failed_record_count,
-      error_messages: resp.records.map {|record| record.error_message }.compact
-    }
-
-    $logger.info("Message sent to #{config[:stream_name]} #{return_message}") if $logger
-
-    return {
-      "code": "200",
-      "message": return_message.to_json
-    }
+    if resp.failed_record_count > 0 
+      return_message = {
+        failures: resp.failed_record_count,
+        failures_data: filter_failures(resp)
+      }
+      $logger.warn("Message sent to #{config[:stream_name]} #{return_message}") if $logger
+    else
+      $logger.info("Message sent to #{config[:stream_name]} successfully") if $logger
+    end
   end
 
   def push_records
-    @batch.each_slice(@batch_size) {|slice| push_batch slice}
-    @batch = []
+    if @records.length > 0 
+      @records.each_slice(@batch_size) do |slice| 
+        push_batch(slice)
+        @batch_count += 1
+      end
+      @records = []
+      @batch_count = 0
+    end
+  end
+
+  def filter_failures(resp)
+    resp.records.filter_map.with_index do |record, i| 
+      avro.decode(@records[i + @batch_size * @batch_count]) if record.key?(:error_message)
```

Comment:
Are we sure that `.key?` is the right method here? as I understand it, `record` is not a hash.